include/onnxruntime/core/framework/execution_provider.h

```diff
@@ -178,6 +178,6 @@ class IExecutionProvider {
   const logging::Logger* logger_ = nullptr;
   // convenience list of the allocators so GetAllocatorList doesn't have to build a new vector each time
   // contains the same instances as allocators_
-  std::vector<gsl::not_null<const IAllocator*>> allocator_list_;
+  std::vector<AllocatorPtr> allocator_list_;
```

Comment:
When we use this list we now want the AllocatorPtr instead of the IAllocator as one layer of indirection was removed vs the usage in ExecutionProviders (where we just looked up the memory info from IAllocator and stored an index number for the EP in order to get the AllocatorPtr from the EP later).

It could be argued our entire usage of AllocatorPtr unnecessarily increments reference counts, so this one off is of no consequence vs the constant creation of new AllocatorPtr instances during execution.

---
In reply to: [443753067](https://github.com/microsoft/onnxruntime/pull/4194#discussion_r443753067) [](ancestors = 443753067)

Reply:
Not sure if I fully understood; I'll check more. Storing 2 shared_ptr objects that point to the same object in the same data structure seems overkill.